### PR TITLE
Adding RegNets to tf.keras.applications

### LINKED ISSUE #15240: Interested in adding RegNets to tf.keras.applications
**System information**.

TensorFlow version (you are using): 2.6.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

RegNets are used in self supervised vision works like SEER. They are also highly efficient and scalable architectures. I intend to train and incorporate RegNet-{X, Y, Z} in `tf.keras.applications`. I think having models that are fast, scalable and efficient be ready to use will greatly help the TF community.

**Will this change the current api? How?**
There will be new handles, namely `tf.keras.applications.RegNet{X, Y, Z}` in the API.

**Who will benefit from this feature?**
The entire TF Keras community.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes

**Current Progress**

I have implemented the building blocks of RegNetY in [AdityaKane2001/regnety](https://github.com/AdityaKane2001/regnety). I have also trained 4 variants of RegNetY on ImageNet-1k and uploaded them to TFHub during Google Summer of Code 2021. 

[Published collection on TFHub](https://tfhub.dev/adityakane2001/collections/regnety/1)
[Repo link](https://github.com/AdityaKane2001/regnety)
[Project report link](https://adityakane2001.github.io/opensource/gsoc2021report)

/cc @sayakpaul

### PR DESCRIPTION
**Pull request by AdityaKane2001**
Please see #15240 and #15419. 

Progress so far:

X variant:

<table>
    <tr>
        <th>Model</th>
        <th>Paper (in %)</th>
        <th>Ours (in %)</th>
        <th>Diff (in %)</th>
        <th>Comments</th>
    </tr>
    <tr>
        <td>X002</td>
        <td>68.9</td>
        <td>67.15</td>
        <td>1.75</td>
        <td>adamw, area_factor=0.25</td>
    </tr>
    <tr>
        <td>X004</td>
        <td>72.6</td>
        <td>71.22</td>
        <td>1.38</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X006</td>
        <td>74.1</td>
        <td>72.37</td>
        <td>1.73</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X008</td>
        <td>75.2</td>
        <td>73.45</td>
        <td>1.75</td>
        <td>adamw, area_factor=0.08</td>
    </tr>
    <tr>
        <td>X016</td>
        <td>77</td>
        <td>75.55</td>
        <td>1.45</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X032</td>
        <td>78.3</td>
        <td>77.09</td>
        <td>1.21</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X040</td>
        <td>78.6</td>
        <td>77.87</td>
        <td>0.73</td>
        <td>adamw, area_factor=0.08, mixup=0.2</td>
    </tr>
    <tr>
        <td>X064</td>
        <td>79.2</td>
        <td>78.22</td>
        <td>0.98</td>
        <td>adamw, area_factor=0.08, mixup=0.3</td>
    </tr>
    <tr>
        <td>X080</td>
        <td>79.3</td>
        <td>78.41</td>
        <td>0.89</td>
        <td>adamw, area_factor=0.08, mixup=0.3</td>
    </tr>
    <tr>
        <td>X120</td>
        <td>79.7</td>
        <td>79.09</td>
        <td>0.61</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
    <tr>
        <td>X160</td>
        <td>80</td>
        <td>79.53</td>
        <td>0.47</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
    <tr>
        <td>X320</td>
        <td>80.5</td>
        <td>80.35</td>
        <td> 0.15</td>
        <td>adamw, area_factor=0.08, mixup=0.4</td>
    </tr>
</table>

Y variant:

<table>
<thead>
  <tr>
    <th></th>
    <th>Paper (in %)</th>
    <th>Ours (in %)</th>
    <th>Diff (in %)</th>
    <th>Comments</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Y002</td>
    <td>70.3</td>
    <td>68.51</td>
    <td>1.79</td>
    <td>adamw, WD=1e-5, area_factor=0.16 mixup=0.2</td>
  </tr>
 <tr>
    <td>Y004</td>
    <td>74.1</td>
    <td>72.11</td>
    <td>1.99</td>
    <td>adamw, area_factor=0.16, mixup=0.2,WD=1e-5</td>
  </tr> 
 <tr>
    <td>Y006</td>
    <td>75.5</td>
    <td>73.52</td>
    <td>1.98</td>
    <td>adamw, area_factor=0.16, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y008</td>
    <td>76.3</td>
    <td>74.48</td>
    <td>1.82</td>
    <td>adamw, area_factor=0.16, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y016</td>
    <td>77.9</td>
    <td>76.95</td>
    <td>0.95</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr> 
 <tr>
    <td>Y032</td>
    <td>78.9</td>
    <td>78.05</td>
    <td>0.85</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr> 
  <tr>
    <td>Y040</td>
    <td>79.4</td>
    <td>78.2</td>
    <td>1.2</td>
    <td>adamw, area_factor=0.08, mixup=0.2</td>
  </tr>
  <tr>
    <td>Y064</td>
    <td>79.9</td>
    <td>78.95</td>
    <td>0.95</td>
    <td>adamw, area_factor=0.08, mixup=0.3</td>
  </tr> 
 <tr>
    <td>Y080</td>
    <td>79.9</td>
    <td>79.11</td>
    <td>0.69</td>
    <td>adamw, area_factor=0.08, mixup=0.3</td>
  </tr>
  <tr>
    <td>Y120</td>
    <td>80.3</td>
    <td>79.45</td>
    <td>0.85</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
  <tr>
    <td>Y160</td>
    <td>80.4</td>
    <td>79.71</td>
    <td>0.69</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
  <tr>
    <td>Y320</td>
    <td>80.9</td>
    <td>80.12</td>
    <td>0.78</td>
    <td>adamw, area_factor=0.08, mixup=0.4</td>
  </tr>
</tbody>
</table>

/cc @fchollet @sayakpaul @qlzh727 
/auto Closes #15240.

### CODE REVIEW

**Comment by innat** [comment]
@AdityaKane2001 
Wondering, looks like it's the first time we're going to have lots of variants of the same model in `tf.keras.applications`. Great job anyway.

**Comment by AdityaKane2001** [comment]
@innat 

Yes, that is the case. Thank you :)

**Comment by MrinalTyagi** [comment]
> @AdityaKane2001 Wondering, looks like it's the first time we're going to have lots of variants of the same model in `tf.keras.applications`. Great job anyway.

@innat Isn't addition of ResNet18 and 34 possible in similar fashion?

**Comment by innat** [comment]
@MrinalTyagi 
I agree. But I don't know why these models aren't there. There're some requests [1](https://github.com/keras-team/keras/issues/15269), [2](https://github.com/keras-team/keras/issues/15270), [3](https://github.com/keras-team/keras/issues/15267), [4](https://github.com/keras-team/keras/issues/15268) are pending. Maybe there're some criteria that I'm unaware of.

File: keras/applications/regnet.py
Comment (lgeiger) [inline keras/applications/regnet.py:647]:
```suggestion
      for i in range(1, depth):
        x = XBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_XBlock_{i}")(x)
```

File: keras/applications/regnet.py
Comment (lgeiger) [inline keras/applications/regnet.py:658]:
```suggestion
      for i in range(1, depth):
        x = YBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_YBlock_{i}")(x)
```

File: keras/applications/regnet.py
Comment (lgeiger) [inline keras/applications/regnet.py:642]: This code already uses f-strings in a few places, how do you feel about using them for things like this as well?
```suggestion
                 name=f"{name}_XBlock_0")(x)
```

File: keras/applications/regnet.py
Comment (lgeiger) [inline keras/applications/regnet.py:664]:
```suggestion
                 name=f"{name}_ZBlock_0")(x)
```

File: keras/applications/regnet.py
Comment (lgeiger) [inline keras/applications/regnet.py:669]:
```suggestion
      for i in range(1, depth):
        x = ZBlock(filters_out,
                   filters_out,
                   group_width,
                   name=f"{name}_ZBlock_{i}")(x)
```

**Review by lgeiger — COMMENTED** [review]
Thanks again for opening this PR! I have two tiny suggestions to slightly improve readability.

**Comment by AdityaKane2001** [comment]
@lgeiger  Always appreciated! Thanks for the changes. I'll merge them tomorrow.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:564]: Line too long. Conversion is...

```
raise ValueError(
          f"Input filters({filters_in}) and output filters({filters_out}) are not equal "
          f"for stride...
```

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:671]: Here as well, consecutive single quote strings.

File: keras/applications/applications_load_weight_test.py
Comment (mattdangerw) [inline keras/applications/applications_load_weight_test.py:74]: I think you need a tf_py_test entry for this test in the BUILD file, you can see example for other load weights tests.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:837]: It looks like this part has not been tested? You are grabbing the last two characters, but your application keys above are all 4 characters long.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:219]: You need to actually use this. See https://github.com/keras-team/keras/blob/6c719b1a2229d554dd7714c22ff7a5a8a84e0ae7/keras/applications/efficientnet_v2.py#L1261 as an example.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:298]: Add an option for this `include_preprocessing=True` see efficientnet_v2 and mobilenet_v3

File: keras/applications/regnet.py
Comment (mattdangerw) [reply keras/applications/regnet.py:564]: Fix other instances in this file as well.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:1549]: To follow out other applications, include `preprocess_input` and `decode_predictions` symbols. `preprocess_input` should just be a passthrough function as you are using rescaling in the model.

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:760]: indent, and add newline after returns block

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:683]: fix indent

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:337]: space before paren

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:516]: remove this comment, the link is in the function docstring already

File: keras/applications/regnet.py
Comment (mattdangerw) [inline keras/applications/regnet.py:597]: remove

**Review by mattdangerw — COMMENTED** [review]
Thanks for the PR! Left a few comments. Do you have the weights for the applications available online somewhere?

File: keras/applications/applications_load_weight_test.py
Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py:74]: I see.

File: keras/applications/regnet.py
Comment (AdityaKane2001) [reply keras/applications/regnet.py:837]: Yes, actually. They are not tested yet. I'm still training the models.

File: keras/applications/regnet.py
Comment (AdityaKane2001) [reply keras/applications/regnet.py:298]: I have added the option:
https://github.com/keras-team/keras/blob/6c719b1a2229d554dd7714c22ff7a5a8a84e0ae7/keras/applications/regnet.py#L803-L805

**Comment by AdityaKane2001** [comment]
@mattdangerw 

Thanks for the review! 

> Do you have the weights for the applications available online somewhere?

I am still training these models, and I'm updating the [tables](https://github.com/keras-team/keras/pull/15702#issue-788479260) at the start of the thread accordingly. I'll test the loading code parallelly.

**Comment by AdityaKane2001** [comment]
@fchollet @mattdangerw 

I have completed training of all the models. I have updated the tables at the [start of the thread](https://github.com/keras-team/keras/pull/15702#issue-788479260) accordingly.  There are a couple of things I want to bring to your notice:

1. `model.predict(X_test)` does not work with grouped convolutions on CPU. For some reason, `model(X_test)` works flawlessly. Thus, I have updated the [`application_load_weight_test.py`](https://github.com/keras-team/keras/pull/15702/files#diff-89b6c8b391af2f95b405f44e3230796aaabf500cad8d49830d2a0fefba86c554R128) file accordingly.  If needed I'll open an issue for the same.
2.  All models are with 2% of the accuracies mentioned in the paper. Larger models are within 1%.

**Comment by AdityaKane2001** [comment]
@fchollet @mattdangerw 

Could you please take a look at this one?

**Comment by mattdangerw** [comment]
@AdityaKane2001 thanks! Do you have the weights hosted online available to download somewhere? We will need that in order to test the model (and eventually upload the weights).

**Comment by AdityaKane2001** [comment]
@mattdangerw 

Yes, I have them in a GCS bucket: https://storage.googleapis.com/ak-regnet-savedmodels/

The `.h5` file paths are as expected, [`~/regnetx008.h5`](https://storage.googleapis.com/ak-regnet-savedmodels/regnetx008.h5) and [`~/regnetx008_notop.h5`](https://storage.googleapis.com/ak-regnet-savedmodels/regnetx008_notop.h5) for example.

**Review by mattdangerw — APPROVED** [review]
Looks great! Approving with two last comments.

Note that we will still need to go through a bit of process on our side to update weights, generate files for the new API and test everything out. If anything comes up there might need to message back here for changes.

Thank you!

**Comment by AdityaKane2001** [comment]
@mattdangerw 

I have made the requested changes. Please run the workflow again.

I wanted to provide performance metrics for all the models as on [keras.io/applications](https://keras.io/api/applications/). However, I am not able to spin up a VM instance of the given specs[^1] on Google Cloud. 

Could you please tell how to go about this?

/cc @sayakpaul 
[^1]: CPU: AMD EPYC Processor (with IBPB) (92 core) - Ram: 1.7T - GPU: Tesla A100

**Comment by AdityaKane2001** [comment]
@fchollet @mattdangerw

Could you please take a look at this one? TIA

**Review by mattdangerw — APPROVED** [review]

**Comment by AdityaKane2001** [comment]
Thanks for the approval!

**Comment by sayakpaul** [comment]
@mattdangerw could you also provide an update on what do we do about this? 

> I wanted to provide performance metrics for all the models as on keras.io/applications. However, I am not able to spin up a VM instance of the given specs1 on Google Cloud.

**Comment by mattdangerw** [comment]
Yeah, re ideal machine for metrics, particularly the performance per step numbers, I am not sure. This is probably a question for @fchollet. We might take a bit to get back to you on this given it's the holidays and a lot of the team is out.

In the mean time, I think we can move ahead trying to land this PR, as the table update will be a separate change to keras.io anyway.

File: keras/applications/applications_load_weight_test.py
Comment (mattdangerw) [inline keras/applications/applications_load_weight_test.py:128]: Is there a reason this needs to be updated? Do things break otherwise? We still run these test cases in TF1 without eager mode enabled, and this line is breaking a number of our tests.

**Review by mattdangerw — COMMENTED** [review]
I've uploaded weights and gen'd api files for this PR, things are looking overall good. However the change to applications_load_weight_tests is breaking right now. Commented on the line. Is that change necessary? Thanks!

File: keras/applications/applications_load_weight_test.py
Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py:128]: @mattdangerw 

Yes, the change is necessary. Grouped convolutions are not yet fully supported on CPUs. We see that `model.predict(X_test)` breaks whereas `model(X_test)` works fine. 

There are a [number of issues](https://github.com/tensorflow/tensorflow/search?q=grouped+convolution&type=issues) discussing this in the TF repo.

Comment (mattdangerw) [reply keras/applications/applications_load_weight_test.py:128]: I think you could trigger the failures by adding a call to `tf.compat.v1.disable_v2_behavior()` after before the call to `tf.test.main` in `applications_load_weight_test`. We can't submit this if we are breaking all these application tests in a TF1 context. We would need to find a change that does not rely on eager mode behavior (.numpy is eager only).

This might mean we need to dig into the difference between direct call vs predict here. It sound like this is an issue with grouped convolutions on CPU that will only appear when compiling a tf.function, is that right?

Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py]: @mattdangerw 

I have found a small solution to this. I have tested the change both with TF2 and using `tf.compat.v1.disable_v2_behavior()` and it works on my end. Could you please take a look and run the workflow again?

https://github.com/keras-team/keras/blob/851ca160f7c38e9effd9e60bcd44b6611073ae60/keras/applications/applications_load_weight_test.py#L127-L132

Comment (mattdangerw) [reply keras/applications/applications_load_weight_test.py]: I will take a look next week! Sorry for the delayed reply, most of the team is out this week. The proposed change would still not submit, because that fallback (the numpy call) would still be run in a TF1 context for the regnet load weights test unless we disable it.

Overall, I think our options are...

 1) Disable the load weights test for regnet (without removing the predict call here), and follow up with a fix.
 2) Fix the underlying CPU/compiled function/grouped convolution issue, and then land this PR.
 3) Work around the bug for regnets somehow (the conversation [here](https://github.com/tensorflow/tensorflow/issues/51825) suggests that using `jit_compile=True` may allow CPU to work, which might give us a way forward).

I would say 3) would be the way to go if we can make it work. We really do want the load weights tests to test the compile predicted function (that's how these will often be used!), and shipping regnets such that predict will fail on CPU by default is not a great out of box experience.

Will follow up next week when people are back in office. Thanks!

Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py]: @mattdangerw 

Thanks for the detailed explanation. 

I guess (2) is not something which can be done in the Keras codebase, as error is thrown in [tensorflow/tensorflow/core/kernels/conv_ops_fused_impl.h](https://github.com/tensorflow/tensorflow/blob/1ef2bbae5f7521c9f1c3a831c162fe9a540df4b1/tensorflow/core/kernels/conv_ops_fused_impl.h#L216-L218).  I'll open an issue in the TF repo regarding this. So I agree that (3) might be the best option.

Lastly, wish you very happy new year!

Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py]: @mattdangerw 

Could you please take a look at this one? TIA

/cc @fchollet @qlzh727

Comment (mattdangerw) [reply keras/applications/applications_load_weight_test.py]: Still here on this! We think we have found a good workaround (option 3), forcing XLA compilation grouped convolutions. https://github.com/keras-team/keras/pull/15868

Once that lands (assuming that doesn't run into road blocks), we can submit this without modifying the predict call in load weights tests.

Comment (AdityaKane2001) [reply keras/applications/applications_load_weight_test.py]: @mattdangerw 

Thanks a lot for this! Really appreciate it.

**Comment by AdityaKane2001** [comment]
@mattdangerw 

Thanks a ton for #15868!

I have tested the code on my end and rolled back https://github.com/keras-team/keras/pull/15702/commits/851ca160f7c38e9effd9e60bcd44b6611073ae60 in https://github.com/keras-team/keras/pull/15702/commits/cf25748f546ef062d49e280cfe757c8d74f23f3e.

**Comment by AdityaKane2001** [comment]
Today these models were pushed to the [official docs](https://www.tensorflow.org/api_docs/python/tf/keras/applications/regnet). I sincerely thank the Keras team for allowing me to add these models. Huge thanks to the [TPU Research Group (TRC)](https://sites.research.google/trc/about/) for providing TPUs for the entire duration of this project, without which this would not have been possible. Thanks a lot to @fchollet for allowing this and guiding me throughout the process. Thanks to @qlzh727 for his guidance in building Keras from source on TPU VMs. Thanks to @mattdangerw for his support regarding grouped convolutions. Special thanks to @lgeiger for his contributions to the code. Last but not least, thanks a ton to @sayakpaul for his continuous guidance and encouragement.

**Comment by mattdangerw** [comment]
Congrats on getting it landed and thanks for all the hard work on this! This is great to have!

**Pull request closed without merging**
